[flang] Lower passing non assumed-rank/size to assumed-ranks

jeanPerier opened the pull request:
Start implementing assumed-rank support as described in https://github.com/llvm/llvm-project/blob/main/flang/docs/AssumedRank.md

This commit holds the minimal support for lowering calls to procedure with assumed-rank arguments where the procedure implementation is done in C.

The case for passing assumed-size to assumed-rank is left TODO since it will be done a change in assumed-size lowering that is better done in another patch.

Implementation of Fortran procedure with assumed-rank arguments is still TODO.

jeanPerier committed:
[flang] Lower passing non assumed-rank/size to assumed-ranks

Start implementing assumed-rank support as described in
https://github.com/llvm/llvm-project/blob/main/flang/docs/AssumedRank.md

This commit holds the minimal support for lowering calls to
procedure with assumed-rank arguments where the procedure implementation
is done in C.

The case for passing assumed-size to assumed-rank is left TODO since it
will be done a change in assumed-size lowering that is better done in
another patch.

Implementation of Fortran procedure with assumed-rank arguments is still
TODO.

test

jeanPerier committed:
fix typo inside assert

clementval commented on flang/lib/Lower/ConvertCall.cpp:
```diff
@@ -1043,39 +1083,55 @@ static PreparedDummyArgument preparePresentUserCallActualArgument(
+  // For ranked actual passed to assumed-rank dummy, the cast to assumed-rank
+  // box is inserted when building the fir.call op. Inserting it here would
+  // cause the fir.if results to be assumed-rank in case of OPTIONAL dummy,
+  // causing extra runtime costs due to the unknwon runtime size of assumed-rank
```
```suggestion
  // causing extra runtime costs due to the unknown runtime size of assumed-rank
```

clementval reviewed (approved):
LGTM

vzakhari commented on flang/lib/Lower/ConvertCall.cpp:
```diff
@@ -942,6 +964,18 @@ static PreparedDummyArgument preparePresentUserCallActualArgument(
 
+  // Create dummy type with actual argument rank when the dummy is an assumed
+  // rank. That way, all the operation to create dummy descriptors are ranked if
+  // the dummy is ranked, which allows simple code generation.
```
I am not sure I am following the comment.  "if the dummy is ranked" seems to contradict the previous condition that "dummy is an assumed rank".  Can you please clarify?

jeanPerier replied:
Indeed, thanks for reading it! updated to "to create dummy descriptors are ranked if the actual argument is ranked"

jeanPerier commented on flang/lib/Lower/ConvertCall.cpp:
```diff
@@ -942,6 +964,18 @@ static PreparedDummyArgument preparePresentUserCallActualArgument(
 
+  // Create dummy type with actual argument rank when the dummy is an assumed
+  // rank. That way, all the operation to create dummy descriptors are ranked if
+  // the dummy is ranked, which allows simple code generation.
```
```suggestion
  // the actual argument is ranked, which allows simple code generation.
```

jeanPerier committed:
Fix typos in comments

Co-authored-by: Valentin Clement (バレンタイン クレメン) <clementval@gmail.com>

jeanPerier committed:
add todo for assumed-rank variables

The current patch only covers the case where the procedure
with assumed rank arguments is not implemented in Fortran.

jeanPerier committed:
handle ignore_tkr(t) case when reboxing is needed

Enforcing the 18.5.3 exposed a weakness in ignore_tkr(t) that could lead
to invalid fir.rebox being generated (using the dummy type was output
which is not the same as the input type which is the type of the actual
argument). Always prepare the dummy  descriptor according to the actual
argument type in case of ignore_tkr(t).
